States/graph.c

```diff
@@ -0,0 +1,203 @@
+#include <stdlib.h>
+#include "graph.h"
+#include <stdio.h>
+#include <math.h>
+#include <stdbool.h>
+#include <string.h>
+
+void setValueTo(int* array, int length, int value) {
+    for (int i = 0; i < length; ++i) {
+        array[i] = value;
+    }
+}
+
+bool allIsVisited(bool* visitedCities, bool* allowedCities, int citiesCount) {
+    for (int i = 0; i < citiesCount; i++) {
+        if (visitedCities[i] == false && allowedCities[i] == true) {
+            return false;
+        }
+    }
+    return true;
+}
+
+void dijkstra(int start, int previousLength, int** adjMatrix, bool* allowedCities, bool* visitedCities, int citiesCount, int* distances, bool* employmentOfCities) {
+    visitedCities[start] = true;
+    int minDistance = INT_MAX;
+    int minCity = -1;
+    for (int end = 0; end < citiesCount; end++) {
+        if (!visitedCities[end] && !employmentOfCities[end]) {
+            if (adjMatrix[start][end] != 0) {
+                int currentDistance = adjMatrix[start][end];
+                if (distances[end] > currentDistance + previousLength) {
+                    distances[end] = currentDistance + previousLength;
+                }
+            }
+        }
+    }
+    for (int i = 0; i < citiesCount; i++) {
+        if (distances[i] != 0 && distances[i] < minDistance & allowedCities[i]) {
+            minDistance = distances[i];
+            minCity = i;
+        }
+    }   
+    if (minCity != -1) {
+        if (!allIsVisited(visitedCities, allowedCities, citiesCount)) {
+            dijkstra(minCity, minDistance, adjMatrix, allowedCities, visitedCities, citiesCount, distances, employmentOfCities);
+        }
+    }
+}
+
+int** createAdjMatrix(int n, int m, Road* roads) {
+    int** matrix = calloc(n, sizeof(int*));
+    if (matrix == NULL) {
+        return NULL;
+    }
+    
+    for (int i = 0; i < n; i++) {
+        matrix[i] = calloc(n, sizeof(int));
+        if (matrix[i] == NULL) {
+            for (int j = 0; j < i; j++) {
+                free(matrix[j]);
+            }
+            free(matrix);
+            return;
+        }
+    }
+    for (int i = 0; i < m; ++i) {
+        Road* currentRoad = roads + i;
```

Comment:
```suggestion
        Road* currentRoad = roads[i];
```
Тем более что арифметика указателей опасна для неокрепших умов и потому запрещена